File: 1129-longest-string-chain/longest-string-chain.cpp
class Solution {
public:
    int longestStrChain(vector<string>& words) {
        int n=words.size();
        vector<int> len(n,1);
        sort(words.begin(),words.end(),[](const string &a,const string &b){
            return a.size() < b.size();
            });
        int ans=0;
        for(int i=0;i<n;i++){
            string str=words[i];
            int temp=0;
            for(int k=0;k<str.size();k++){
                string s=str;
                str.erase(k,1);
                
                for(int j=0;j<i;j++){
                    if(words[j]==str){
                        temp=max(temp,1+len[j]);
                    }
                }
                str=s;
            }
            len[i]=max(len[i],temp);
            ans=max(ans,len[i]);
        }
        return ans;
    }
};

File: 1129-longest-string-chain/longest-string-chain.py
class Solution:
    def longestStrChain(self, words: list[str]) -> int:
        n=len(words)
        words.sort(key=len)
        dp=[1]*n
        parent=[-1]*n
        max_len=1
        for i in range(n):
            parent[i]=i
        for i in range(n):
            word_a=words[i]
            for k in range(len(word_a)):
                prev_word=word_a[:k] + word_a[k+1:]
                for j in range(i):
                    if words[j]==prev_word:
                        if dp[j]+1 > dp[i]:
                            dp[i]=dp[j]+1
                            max_len=max(max_len,dp[i])
        return max_len
                


        